### DISCUSSION TITLE
SSE

### DISCUSSION THREAD

**Discussion by manson** [kamalshkeir/kago#15] [General]
Hi!
I'm here with another issue. I've been playing lately with SSE and used you example:
```
// will be hitted every 1-2 sec, you can check anything if change and send data on the fly using c.StreamResponse
app.SSE("/sse/logs",func(c *kamux.Context) {
lenStream := len(logger.StreamLogs)
if lenStream > 0 {
	lastOne := lenStream-1
	err := c.StreamResponse(logger.StreamLogs[lastOne])
	if err == nil{
		logger.StreamLogs=[]string{}
	}
}
})
```
1) It's a bit old: now instead of StreamResponse the Stream is used. Now it looks like this:
```
sbus.App.SSE("/sse/logs",func(c *kmux.Context) {		
		lenStream := len(logger.StreamLogs)
		if lenStream > 0 {
			lastOne := lenStream-1
			lastLog := logger.StreamLogs[lastOne]
			c.Stream(lastLog)
                        return
		}
		c.Stream("")
	})
```
2) Logs in example could be lost if they appear faster than handler run period, but ok, this is only an example, but should be noted.
3) It's a bit strange way to periodically call sse handler - never met like this before, but ok, I even found this useful to make not direct call to Stream right away when data appeared but in time, filling a queue and later (1-3 seconds ok) send back. But exactly this example won't work - it will rise an exception on client side ```EventSource's response has a MIME type ("text/plain") that is not "text/event-stream". Aborting the connection.```. This is if no logs exist. But if I put c.Stream("") in the end it will work ok and data will come to client, but if no data, it will send to client empty strings. This is not good. Wrong. Stream should result to client only when I want, when data exists, but not periodically. I.e. It's ok to periodically call a handler, but only I should decide whether to send back data or not in it.

Theoretically I could workaround this like in WS example - storing connection of caller and later send to stream, but in this case I cannot off this periodical handler calls. Besides, I suppose, you are already storing these connections to process periodic handler and additional copy I could do is not good.

So I think there needs to be made next changes:
1) Ability to off and on periodic sse handler call - sometimes it is useful, but sometimes not needed. Maybe 2 handlers? Usual and periodic?
2) Remove client-side exception if handler does not send back data to the stream during handler call. I don't exactly remember, but SSE is simple only at first glance - it has its internal mechanism to keep connection alive (reestablish connections, internal pings etc.), it checks if data lost during connection break by monitoring packet ID (on connection client sends last packet ID it received, and server resends packets starting from that id, etc.). 

Without this It looks like SSE not much useful...

**Comment by kamalshkeir** [comment]
@manson Hi bro, 
indeed SSE is less powerful than websockets, for instance sse can handle only 5 connections i think, and it's not me who is calling the handler, it is called every 1-2 seconds by the browser automatically (not sure but firefox also may have some different behavior like calling the endpoint more or less frequentily), this is how sse protocol is implemented, so it's like browser pooling for and endpoint that you cannot stop. 
I think there is nothing you can do with SSE that you cannot do with a simple pooling or websockets, and it is used for push notifications usually, and overall my bus implemantation can make handling these ws connections way easier using the python, js, and go client in a pubsub style

**Reply by manson** [reply]
Strange. I've used SSE server implementation in golang in my other project with Echo, it called handler only once on establishing connection. All other reconnections was hidden from my code. 

```sse can handle only 5 connections``` - this is for domain. Usually I don't have more than 1 sse connection to a server and don't have more than 2 opened tabs with the same domain. It is useless to do like this. Anyway - I don't care about this.

```and it's not me who is calling the handler``` - it is bad. Anyway, in my experience with SSE before, I did not have to send empty strings to client. 

```I think there is nothing you can do with SSE that you cannot do with a simple pooling or websockets,``` - simple pooling is too simple, sse has some keepalive logic and control of lost packets. Websockets is full duplex protocol, it is heavier and consumes more memory. I consider this is good for serverside communications, but not for client-server in my usecases. It is good for specific usage. In all my usecases and as I think in all common usecases websockets is too much. Besides, when I used it in nodejs through tls it made nodejs use a lot of CPU for crypting purposes. For go of couse this is not critical (or using reverse proxy in the middle) - just saying my experience. For SSE this all is somehow less heavy. And I don't want to use two channels to send commands to the server - http and websocket, but if not - why should I keep extra possibilities, making them anyway idle?

```and it is used for push notifications usually``` - exactly what I need. And in a regular website only notifications are needed, not more. For last years I needed websockets exactly zero times. All other times I used it - I used it only because my tools or frameworks used it, but I could do the same with less efforts or resources... For example I had a big project in MeteorJs that has a websocket internal client-server implementation as a core engine, but after all with much pain had to remove this core engine, based on websockets, to overcome limits during high load and go back to a regular HTTP calls and SSE notifications...

I used this lib: ```github.com/alexandrevicenzi/go-sse``` without these problems. It is a bit old, maybe there are more modern libs, did not search for a couple of years (now I would search if new appeared). Anyway, if you think this is useless for you to reimplement, I will try to attach this lib for my use cases or similar, because I cannot use your SSE implementation.

**Reply by kamalshkeir** [reply]
@manson it should work now, update korm to v1.5.7

**Reply by kamalshkeir** [reply]
i try it before , and indeed i got this error `EventSource's response has a MIME type ("text/plain") that is not "text/event-stream". Aborting the connection.` when no response was returned from the handler, now , no response returned unless you push using c.Stream

**Reply by kamalshkeir** [reply]
`and it is used for push notifications usually` what i want to say here , is that using SSE here for push notifications at least for me is not easier and not sure it is more efficient , because subscribing from clientJS on topic for notification and from the server i can publish on this topic whenever i want when some actions happen, rather than checking every X time if something changed then showing notifications, also combining SSE or bus for notifications with the service worker from PWA is very convenient if you are motivated to try, this will allow you to send notifications offline, and fetch data when connection is back which is called sync , i will make a video on PWA in the futur, but if you want to start using it with korm, you can include this tag in any template html file `{{template "pwa"}}` like here https://github.com/kamalshkeir/korm-dashboard/blob/master/templates/admin/admin_all_models.html#L158, this will init the PWA using assets/static/sw.js and assets/static/manifest.json.

you can communicate between service worker and you js files or html like so:


```js
// app.js - Somewhere in your web app
navigator.serviceWorker.controller.postMessage({
  type: 'MESSAGE_IDENTIFIER',
});
// sw.js
// On the Service Worker side we have to listen to the message event
self.addEventListener('message', (event) => {
  if (event.data && event.data.type === 'MESSAGE_IDENTIFIER') {
    // do something
  }
})
```

**Reply by kamalshkeir** [reply]
@manson  another fix v1.5.8 for SSE, i was not flushing the response writer after each call of c.Stream, so golang by default add them to a buffer , for example : 


```go
app.SSE("/test",func(c *kmux.Context) {
    for i := 0; i < 5 ;i++ {
        c.Stream("hello")
        time.Sleep(time.Second)
    }
})
```

this was producing before : 5 hello at the same time when handler return

now it will send 1 hello each second , not all of them after 5 second

**Reply by manson** [reply]
```using SSE here for push notifications at least for me is not easier and not sure it is more efficient , because subscribing from clientJS on topic for notification and from the server i can publish on this topic whenever i want when some actions happen, rather than checking every X time if something changed then showing notifications``` - what happen if I store kmux.Context for futher usage and use c.Stream from some outer functions? Because in theory basically SSE is not something different that Websockets except it only sends data to client (I know, technically implementations may be totally different).  In 99% situations exactly this functionality is necessary, not full duplex communication to a server. Besides, for SSE from client-side I can use the simple EventListener without any external libraries to consume the notifications. And 3d objection - by using that library I noted above, I could use it just like a simple channel to a client without any difficulties. So I never checked data changes, I just did c.Stream (in your notation) at any point of application logic and send it to a client. You just mix technical implementation and logical. From my point of view (as a application developer, not system developer) SSE is a light version of websockets half duplex for notifications only - from server to client. And usage suppose not differ much from using websocets

**Reply by manson** [reply]
> @manson another fix v1.5.8 for SSE, i was not flushing the response writer after each call of c.Stream, so golang by default add them to a buffer , for example :
> 
> ```go
> app.SSE("/test",func(c *kmux.Context) {
>     for i := 0; i < 5 ;i++ {
>         c.Stream("hello")
>         time.Sleep(time.Second)
>     }
> })
> ```
> 
> this was producing before : 5 hello at the same time when handler return
> 
> now it will send 1 hello each second , not all of them after 5 second

Ok, thanks. Will try it tonight. Don't know why this occurs every 1 second because after getting data client makes new connection to a server right away, gets new data and so on and all this should looks like quick and immediate

**Reply by manson** [reply]
```PWA is very convenient if you are motivated to try``` - currently I am far from it. I just step by step tries to implement an application a planned and of course it doesn't have all the framework KORM features :-)

**Reply by manson** [reply]
Perhaps I will eventually come to use web sockets in this application if the work of SEE does not satisfy me, but for now I want to squeeze everything I can out of SSE

**Reply by kamalshkeir** [reply]
> what happen if I store kmux.Context for futher usage and use c.Stream from some outer functions? 

i think it should work like this, because c.Stream write to the response writer which is inside the context, you can also do it like in this example using a channel : https://dev.to/rafaelgfirmino/golang-and-sse-3l56

```go
var SSEChan = make(chan string,10)
app.SSE("/test",func(c *kmux.Context) {
   timeout := time.After(1 * time.Second)
   select {
       case ev := <-SSEChan:
           c.Stream(ev)
           // or
           // fmt.Fprintf(c.ResponseWriter, "data: %v\n\n", ev)  //same as c.Stream
       case <-timeout:
           fmt.Fprintf(w, ": nothing to sent\n\n")
    }
})
```

**Reply by manson** [reply]
Hmmm... Very interesting. Simple and looks like workable. And as far as I understand quite reliable due to usage of channel that keeps data before send... Thanks! Will try this tonight

**Reply by manson** [reply]
@kamalshkeir 
Hi! Sorry, I was away for a time being. Just wanted to try a simple example you supplied. This one:
```
sbus.App.SSE("/test",func(c *kmux.Context) {
    for i := 0; i < 5 ;i++ {
        c.Stream("hello")
        time.Sleep(time.Second)
    }
})
```
it looks like after first  c.Stream("hello") the StreamWriter becomes closed. So here I have 4 errors in console:
``` [ERROR] github.com/kamalshkeir/kmux.(*Context).Stream [line:190] : flate: closed writer```
And client gets only the first message :-(
I have updated all before:
```
	github.com/kamalshkeir/kago v1.2.7
	github.com/kamalshkeir/kenv v1.0.0
	github.com/kamalshkeir/kmux v1.7.8
	github.com/kamalshkeir/korm v1.6.0
	github.com/kamalshkeir/ksbus v1.0.5
```

**Reply by kamalshkeir** [reply]
hello @manson , i got the same error, so fixed  --> v1.6.2
it was because i enabled Gzip by default, so the error is fixed, if you try again with Gzip, but you will receive the 5 messages at the same time, gzip don't let sse flush the response before it finish

**Reply by manson** [reply]
@kamalshkeir 
Thanks! now it works as expected.

**Reply by kamalshkeir** [reply]
@manson with the last version it worked also with Gzip middleware as expected, it was missing a check in the middleware if the request is SSE it should not be gziped, but still not used by default, you can add it using `srvBus.App.Use(kmux.Gzip())`

Also, Big news, almost 10 hours working on swagger documentation, i cannot wait to show you how easy it is

**Reply by manson** [reply]
@kamalshkeir 
After update (dont know which one) stoped working route mask ```/test/id:string``` if id is like uuid (123-324-fds-343). But works with this id if it is ""/test/id:slug". Did you change the rules? It was "string" to accept all data here but now not...

**Reply by kamalshkeir** [reply]
> @kamalshkeir 
> After update (dont know which one) stoped working route mask ```/test/id:string``` if id is like uuid (123-324-fds-343). But works with this id if it is ""/test/id:slug". Did you change the rules? It was "string" to accept all data here but now not...

the problem here is there is no string choice, it's 'str' instead, so before when you enter something not handled it was  defaulted to slug , but i changed to str, that's why string not working anymore, it wasn't a choice before, so if you want slug you should keep it as slug

**Reply by kamalshkeir** [reply]
@manson https://www.youtube.com/watch?v=RupARTkPzf4

**Reply by manson** [reply]
``` there is no string choice``` - Ok, maybe I've missed something. I thought it existed. Now I see, thanks!

**Reply by manson** [reply]
> @manson https://www.youtube.com/watch?v=RupARTkPzf4

Very, very useful feature! And simple to use. Very well. I've subscribed to your youtube channel for future videos :-)

**Reply by kamalshkeir** [reply]
@manson https://www.youtube.com/watch?v=DQcySQW0r58, small update, but make it better, hope you like it

**Reply by manson** [reply]
> @manson https://www.youtube.com/watch?v=DQcySQW0r58, small update, but make it better, hope you like it

Sure I like it. It is getting more and more close to perfection! :-) Of cause IMHO, but I initially liked your coding style (the same like mine) and ways to implement all this (very close as I would do). but I'm lazy to do like this, and I'm more an application developer (uses the existing libs and code to compose my application) but you in this sense are a system developer (not low-lever or OS, but rather a developer of bricks - libs and frameworks in my vision). For example I like UI in pocketbase I met before your Kago. but didn't like the way it implemented and the internals, so even did not start playing with it. So, I'm glad I met your project and possibly gave some usefull ideas to you :-)

**Reply by kamalshkeir** [reply]
> > @manson https://www.youtube.com/watch?v=DQcySQW0r58, small update, but make it better, hope you like it
> 
> Sure I like it. It is getting more and more close to perfection! :-) Of cause IMHO, but I initially liked your coding style (the same like mine) and ways to implement all this (very close as I would do). but I'm lazy to do like this, and I'm more an application developer (uses the existing libs and code to compose my application) but you in this sense are a system developer (not low-lever or OS, but rather a developer of bricks - libs and frameworks in my vision). For example I like UI in pocketbase I met before your Kago. but didn't like the way it implemented and the internals, so even did not start playing with it. So, I'm glad I met your project and possibly gave some usefull ideas to you :-)

Some useful ideas ?, no, all of this wouldn't be that beautiful without your ideas, you are the biggest contributor for sure 😂

**Reply by kamalshkeir** [reply]
> > > @manson https://www.youtube.com/watch?v=DQcySQW0r58, small update, but make it better, hope you like it
> > 
> > Sure I like it. It is getting more and more close to perfection! :-) Of cause IMHO, but I initially liked your coding style (the same like mine) and ways to implement all this (very close as I would do). but I'm lazy to do like this, and I'm more an application developer (uses the existing libs and code to compose my application) but you in this sense are a system developer (not low-lever or OS, but rather a developer of bricks - libs and frameworks in my vision). For example I like UI in pocketbase I met before your Kago. but didn't like the way it implemented and the internals, so even did not start playing with it. So, I'm glad I met your project and possibly gave some usefull ideas to you :-)
> 
> Some useful ideas ?, no, all of this wouldn't be that beautiful without your ideas, you are the biggest contributor for sure  

btw , do some push on korm, you should be in the contributors list @manson

**Reply by manson** [reply]
Another thought for you. As Kago itself became useless because of splitting it to modules I suggest you to rewrite it and use as a starting point of an application if user wish to use it (besides that he can use modules separate as it is now). Say, users whould import this as it was at the very beginning, but user will get access to all its internals (kmux, korm, klog etc) from this. Say, somethig like:
```
app := kago.NewServer().WithDashboard().WithShell();
router ;= kago.GetRouter() // returns instance of kmux
orm := kago.GetOrm(SQLITE) // or MONGO or others; returns korm or mongkorm
log := kago.GetLogger()
```
etc. This is for a common use when developer understans that he will be using all the possiblities of Kago that imports all other libs, or for simplifying the start of using this. Because different libs for composing user's application intially a little confusing - is this a solid framework or a collection of different libs? In my view, although user can use it as separate libs, IT IS a solid single framework. 

Besides in future the main Kago may become (who knows) a central "generator" of some predefined infrastructure. For example, if I llike to create two nodes for a database (or events) syncronization (horizontal sharding) I would use somethis like:
```
node := Kago.NewNode(env.NODE_TYPE) // in env NODE_TYPE=slave or master
if env.NODE_TYPE == nodes.SLAVE {
    node.connect(env.MASTER_URL) // or node.DiscoverMaster() or node.Register(env.DNS, client) 
}
node.SyncMode = modes.MASTER_SLAVE // or MASTER_MASTER
router := node.Router()
```
(or nodes may be equal but run elections to select master or without master at all - as user needs)
And this automatically does some internal preliminary code, in other case user would implement of his own.
Or integrate some additional functionality into Kago like backuping and restoring
```
s:= kago.NewServer()
job := s.backup.NewJob(backups.FTP, url, user, password) // or aws object store, or file copy or whatever
job.route = "/backup/download" // to download it from client side
job.job_route = "/backup/job" // to start dowbload from client side
job.schedule = "1 * * * * " // schedule backups
job.jobs_route = "/backup/list" // get the list of downloads
job.router = s.GetRouter()
```
This is just a scketch
Or create something more complex.
In this case we have a solid framework and a separate modules one can use as he likes...

But all this is just to imaginge the step further :-)

**Reply by manson** [reply]
> btw , do some push on korm, you should be in the contributors list @manson

I doubt I will do code contributions directly - no time and too much lazyness :-)  But to give some ideas - as being sayid here where I live: "talking is not carrying bags" :-)

**Reply by kamalshkeir** [reply]
lot of good stuffs

**Reply by kamalshkeir** [reply]
@manson https://www.youtube.com/watch?v=r7rbMrTkVek , i'm a machine 😊

**Reply by manson** [reply]
> @manson https://www.youtube.com/watch?v=r7rbMrTkVek , i'm a machine blush

You're a developer-monster! :-)

**Reply by manson** [reply]
> > @kamalshkeir
> the problem here is there is no string choice, it's 'str' instead, so before when you enter something not handled it was defaulted to slug , but i changed to str, that's why string not working anymore, it wasn't a choice before, so if you want slug you should keep it as slug

Sorry to returning to this point, but I cannot get it work so that parameter gets any param. If I set it to "id:str" I expect to get in Id parameter whatever came. But if this is a string with dash ("111-222") it stops working (not found route). I don't know what param could be in advance, so wish to process it later on my own. And I don't want to set 2 params, one for "str" and one for "slug". Besides, who knows what another format it could skip/not found if param won't fit its masks, for example if I have autogenerated param for some reason....

**Reply by kamalshkeir** [reply]
@manson hi bro.
The most generic type parameter is slug : 

```go
func WrapRegexParam(name, typeToValidate string) string {
	var wrappedString string
	switch typeToValidate {
	case "str", "string":
		wrappedString = `(?P<` + name + `>\w+)`
	case "int", "num":
		wrappedString = `(?P<` + name + `>\d+)`
	case "slug":
		wrappedString = `(?P<` + name + `>[a-z0-9]+(?:-[a-z0-9]+)*)`
	case "float":
		wrappedString = `(?P<` + name + `>[-+]?([0-9]*\.[0-9]+|[0-9]+))`
	default:
		wrappedString = `(?P<` + name + `>[a-z0-9]+(?:-[a-z0-9]+)*)`
	}
	return wrappedString
}
```

so this example will work with any parameter type, even with "111-222" or "abc-123" : 

```go
app.GET("/param:slug", func(c *kmux.Context) {
		fmt.Println(c.Param("param"))
		c.Text("ok")
	})
```

maybe i'm not understanding your problem, if so, give me an example

**Reply by kamalshkeir** [reply]
@manson 

updated version of kmux : 

```go
func WrapRegexParam(name, typeToValidate string) string {
	var wrappedString string
	switch typeToValidate {
	case "int":
		wrappedString = `(?P<` + name + `>\d+(?:-\d+)*)` //  '/name:int' can handle /2 /2-3 /2-3-4
	case "float":
		wrappedString = `(?P<` + name + `>[+-]?([0-9]*[.])?[0-9]+(?:[+-]?([0-9]*[.])?[0-9]+)*)` //  '/name:float' can handle /2.2 /2.2-3.3 /2.2-3.343-4
	case "any":
		wrappedString = `(?P<` + name + `>[^\/]+)` //  '/name:any' can handle all previous and in this case /favicon.ico
	default:
		wrappedString = `(?P<` + name + `>[a-z0-9]+(?:-[a-z0-9]+)*)` //  '/:name' no more str or string or slug, the default now is slug so it will work with /abc-123 but not with /favicon.ico and /abc-2.3
	}
	return wrappedString
}
```

this will not break previous use of /name:str or others
you can execute go get -u to update kmux

**Reply by manson** [reply]
@kamalshkeir 
Thanks. I needed "any" as a parameter. Now it works in my routes, but admin part is broken - when entering to tables, it says "not found" :-(

**Reply by kamalshkeir** [reply]
> @kamalshkeir 
> Thanks. I needed "any" as a parameter. Now it works in my routes, but admin part is broken - when entering to tables, it says "not found" :-(

ah, thanks, i will check it

**Reply by kamalshkeir** [reply]
@manson , fixed, go get -u , kmux is at version v1.9.2

**Reply by manson** [reply]
> @manson , fixed, go get -u , kmux is at version v1.9.2

Thanks! Looks like it works now

**Comment by manson** [comment]
```maybe i'm not understanding your problem, if so, give me an example``` 
I needs param to be "abc" and "123-234" at the same route. I.e. - no limits. Well, letters, digits, and some meaningful chars (.-+^$#_=!@)  Well, why cant it just be splitted by "/"? Say, if no type supplied "/:id" or "/id:any" so in this case just split by "/" and pass like that without regex?
